File: ui/pages/cart_page.py
from playwright.sync_api import Locator, Page

from pages.base_page import BasePage


class CartPage(BasePage):
    def __init__(self, page: Page, base_url: str) -> None:
        super().__init__(page, base_url)
        self._cart_items: Locator = page.locator(".cart_item")
        self._continue_shopping: Locator = page.locator("#continue-shopping")
        self._checkout: Locator = page.locator("#checkout")
        self._cart_badge: Locator = page.locator(".shopping_cart_badge")

    def expect_loaded(self) -> None:
        self.page.wait_for_url("**/cart.html")

    def remove_item_by_name(self, product_name: str) -> None:
        item = self._cart_items.filter(has=self.page.get_by_text(product_name, exact=True))
        item.get_by_role("button", name="Remove").click()

    def continue_shopping(self) -> None:
        self._continue_shopping.click()

    def proceed_to_checkout(self) -> None:
        self._checkout.click()

    @property
    def cart_items(self) -> Locator:
        return self._cart_items

    @property
    def cart_badge(self) -> Locator:
        return self._cart_badge

File: java/playwright/src/test/java/com/tnd/playwright/pages/CartPage.java
package com.tnd.playwright.pages;

import com.microsoft.playwright.Page;

/**
 * Page object for SauceDemo cart screen (Playwright Java).
 *
 * <p>Purpose: encapsulate cart-level actions required before checkout.
 */
public class CartPage extends BasePage {
    /**
     * Creates cart page object bound to active Playwright page.
     */
    public CartPage(Page page, String baseUrl) {
        super(page, baseUrl);
    }

    /**
     * Asserts route transition to cart page.
     */
    public void expectLoaded() {
        page.waitForURL("**/cart.html");
    }

    /**
     * Moves from cart page to checkout step one.
     */
    public void checkout() {
        page.locator("#checkout").click();
        page.waitForURL("**/checkout-step-one.html");
    }
}
